paas_app_charmer/rabbitmq.py

```diff
@@ -0,0 +1,216 @@
+
+
+class RabbitMQRequires(Object):
+    """RabbitMQRequires class.
+
+    Attributes:
+        on: ObjectEvents for RabbitMQRequires
+        port: amqp port
+    """
+
+    on = RabbitMQServerEvents()
+    port = 5672
+
+    def __init__(self, charm: CharmBase, relation_name: str, username: str, vhost: str):
+        """Initialize the instance.
+
+        Args:
+           charm: charm that uses the library
+           relation_name: name of the RabbitMQ relation
+           username: username to use for RabbitMQ
+           vhost: virtual host to use for RabbitMQ
+        """
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.username = username
+        self.vhost = vhost
+        self.framework.observe(
+            self.charm.on[relation_name].relation_joined,
+            self._on_rabbitmq_relation_joined,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_changed,
+            self._on_rabbitmq_relation_changed,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_departed,
+            self._on_rabbitmq_relation_changed,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_broken,
+            self._on_rabbitmq_relation_broken,
+        )
+
+    def _on_rabbitmq_relation_joined(self, _: HookEvent) -> None:
+        """Handle RabbitMQ joined."""
+        self.on.connected.emit()
+        self.request_access(self.username, self.vhost)
+
+    def _on_rabbitmq_relation_changed(self, _: HookEvent) -> None:
+        """Handle RabbitMQ changed."""
+        if self.rabbitmq_uri():
+            self.on.ready.emit()
+
+    def _on_rabbitmq_relation_broken(self, _: HookEvent) -> None:
+        """Handle RabbitMQ broken."""
+        self.on.departed.emit()
+
+    @property
+    def _rabbitmq_rel(self) -> Relation | None:
+        """The RabbitMQ relation."""
+        return self.framework.model.get_relation(self.relation_name)
+
+    def rabbitmq_uri(self) -> str | None:
+        """Return RabbitMQ urs with the data in the relation.
+
+        It will try to use the format in rabbitmq-k8s or rabbitmq-server.
+        If there is no relation or the data is not complete, it returns None.
+
+        Returns:
+            The parameters for RabbitMQ or None.
+        """
+        rabbitmq_k8s_params = self._rabbitmq_k8s_uri()
+        if rabbitmq_k8s_params:
+            return rabbitmq_k8s_params
+
+        # rabbitmq-server parameters or None.
+        return self._rabbitmq_server_uri()
+
+    def request_access(self, username: str, vhost: str) -> None:
+        """Request access to the RabbitMQ server.
+
+        Args:
+           username: username requested for RabbitMQ
+           vhost: virtual host requested for RabbitMQ
+        """
+        if self.model.unit.is_leader():
+            if self._rabbitmq_rel:
```

Comment:
minor: what about using same pattern as the other method?
"if not self._rabbitmq_rel
return None"